pkg/distributor/push.go

```diff
@@ -78,19 +77,37 @@ func (cfg *RetryConfig) Validate() error {
 // Handler is a http.Handler which accepts WriteRequests.
 func Handler(
 	maxRecvMsgSize int,
+	requestBufferPool util.Pool,
 	sourceIPs *middleware.SourceIPExtractor,
 	allowSkipLabelNameValidation bool,
 	limits *validation.Overrides,
 	retryCfg RetryConfig,
 	push PushFunc,
+	reg prometheus.Registerer,
 	logger log.Logger,
 ) http.Handler {
-	return handler(maxRecvMsgSize, sourceIPs, allowSkipLabelNameValidation, limits, retryCfg, push, logger, func(ctx context.Context, r *http.Request, maxRecvMsgSize int, buffers *util.RequestBuffers, req *mimirpb.PreallocWriteRequest, _ log.Logger) error {
-		err := util.ParseProtoReader(ctx, r.Body, int(r.ContentLength), maxRecvMsgSize, buffers, req, util.RawSnappy)
+
+	uncompressedBodySize := promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
+		Name:    "cortex_distributor_uncompressed_request_body_size_bytes",
+		Help:    "Size of uncompressed request body in bytes.",
+		Buckets: uncompressedBodySizeBuckets,
+	}, []string{"user"})
```

Comment:
Could this be a native histogram?

Reply:
good idea. 👍

Reply:
_changed @ b8b796ed9550820bae0392ed8aac138a05f13886_